### ISSUE TITLE
Q: Firefox 54?

### ISSUE THREAD

**Issue by congma** [willsALMANJ/pentadactyl-signed#3]
Hi,

Is there news on whether PentadactylSigned can (or will) work with Firefox 54?

Thanks!

**Comment by jimogsh** [comment]
Will Pentadactyl stop support for Firefox after version 58?

**Comment by polyzen** [comment]
@jimogsh, I forget which release will drop XUL and XPCOM.

I have heard the main developer has been working on API for Firefox, but no news on Penta being ported to WebExtensions: 5digits/dactyl/issues/82

Seems efforts are being focused on a Vimium port: philc/vimium/issues/2425
Here's the Vimperator WE port tracker (which seems to point to Vimium's): vimperator/vimperator-labs/issues/705

**Comment by jimogsh** [comment]
@polyzen Thx a lot.

**Comment by wshanks** [comment]
@congma Have you tried it? I haven't used Pentadactyl in a while, so I'm not sure if there have been breaking changes. My general approach with this repo is to just pull master from https://github.com/5digits/dactyl, but I have temporarily switched to a fork a couple times in order to get compatibility fixes before they are merged into master. I don't see any recent new issues on https://github.com/5digits/dactyl right now.

@jimogsh The most recent plan I know about is described [here](https://blog.mozilla.org/addons/2017/02/16/the-road-to-firefox-57-compatibility-milestones/comment-page-1/). That says that XUL (and thus Pentadactyl) will not work in Firefox 57, so earlier than after 58.

@polyzen Ugh, thanks for the links. I was more active in trying to find a replacement for Pentadactyl last year. I tried posting to the Vimium issue tracker [last November](https://github.com/philc/vimium/issues/1048#issuecomment-261846071) but got no response so assumed there was no interest from the Vimium developers. I made some attempts to port Vimium myself back then but there were too many WebExtensions incompatibilities to work through given the amount of time I had to work on it.

Then I found Simon Lydell's work on adding keyboard support to WebExtensions in this [Bugzilla ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1215061#c23)  and https://github.com/lydell/webextension-keyboard. I took his comments of

> I'm interested in making VimFx a WebExtension. I'm fully aware that we might not be able to port _everything,_ and that's totally fine as long we can just get the basic features going.

and

> TL;DR: Turns out not much at all needs to be done – for VimFx's purposes (I'm dropping all other ambitions). What's needed is: One bug fix, and one new shortcut for the Commands API. No new APIs.

to mean that a VimFx port was in the works and would not be too difficult. So I switched from Pentadactyl to VimFx (I wanted to switch away from Pentadactyl any way so I could use multiprocess mode in Firefox). I had actually grown to like VimFx quite a lot -- not quite as smooth as Pentadactyl but still very usable. Now I checked in on the progress and found that all efforts to port VimFx have been abandoned: https://github.com/akhodakivskiy/VimFx/issues/860 It looks like I have some catching up to do on all of the Vimium progress. I don't know the Chrome extensions very well. My impression was that cVim had a nicer feature set than Vimium, but I guess I will take whatever I can get at this point.

**Comment by polyzen** [comment]
I have been using VimFx for some time.. Totally forgot to mention its progress :p. 

Vimium was nice during my casual Chrom use.

There's also https://github.com/lusakasa/saka-key in the works.

**Comment by congma** [comment]
@willsALMANJ I mostly use Firefox ESR.  I tried 54, and pentadactyl installs fine, is shown as enabled in the list of addons, but it doesn't actually run.  I hadn't seen anything like that before.

**Comment by iegorka** [comment]
Just for info. There is this problem with the plugin loading:

```
pentadactyl-signed: bootstrap: init
pentadactyl-signed: [Begin base]
pentadactyl-signed: Loading base: TypeError: /^function .*?\((.*?)\)/.exec(...) is null

pentadactyl-signed: bootstrap: TypeError: /^function .*?\((.*?)\)/.exec(...) is null
Class@resource://dactyl/bootstrap.jsm -> resource://dactyl/base.jsm:927:1
@resource://dactyl/bootstrap.jsm -> resource://dactyl/base.jsm:1817:12
loadSubScript@resource://dactyl/bootstrap.jsm:26:12
```

**Comment by wshanks** [comment]
I think it would be best to open an issue on https://www.github.com/5digits/dactyl about this. We can update the xpi here once a patch is either pushed there or at least a working pull request is opened. Looking at the developer notes for Firefox 54, it doesn't seem llike there were many significant compatibility changes. There is one note about a change to `\b` in regular expressions that might be related (https://bugzilla.mozilla.org/show_bug.cgi?id=1338373). The lines in question seem to be doing something very complicated with a regular expression -- creating a custom class inside of a custom class constructor by `eval`ing the result of four nested string replacements with a regular expression and some ternary operators mixed. Not exaclty the most readable for debugging purproses.

**Comment by congma** [comment]
Thanks everyone here and in related posts.

The main pentadactyl repository looks dormant (again).  If there's going to be no updates or merges there, will you rebuild and release using your own branch of pentadactyl, @willsALMANJ ?

**Comment by wshanks** [comment]
Here is a build with my fx54 branch: https://github.com/willsALMANJ/pentadactyl-signed/releases/tag/7308a

I didn't set it to auto-update. You can try it out and see how it does. In my quick testing, it seemed like there were still other problems that needed to be fixed in 54 (like no link hint labels) but I didn't have time to look into them.

@congma Are you still planning to use ESR or will you update to 54 completely? It would be good to test the new build on some version of Firefox before 54 to make sure it works okay there (I think it should). I can set it to auto-update in that case.

**Comment by congma** [comment]
@willsALMANJ, thank you.  For the long term, I'm at loss like everyone.  For now, I'll give 54 a try later.

**Comment by wshanks** [comment]
I tried out the patched xpi in ESR and it seemed to work okay so I pushed it into the auto-update channel. It still seemed buggy to me in Firefox 54. If someone else posts more patches I will make a new release. I poked around at it a little bit but it is hard to justify the time given that I don't use Pentadactyl any more. I think it would not be too bad to debug the problems though -- just look at the errors in the error console, put break points at those points in the addon debugger in both Firefox 53 and Firefox 54 and see what looks different. There aren't many clues in the developer notes for 54 -- I would guess that most of the Mozilla people have already stopped paying attention to what is breaking for legacy addons.

**Comment by wshanks** [comment]
Just to make sure the timing is clear -- Firefox 57 is scheduled to be released on November 14, 2017 and Pentadactyl will definitely not work with it. Firefox 52 ESR will be supported until June 26, 2018. So you could try to patch the bugs with 54 now and whatever extra bugs are added with 55 and 56, but after five months you'll have to drop back to 52 any way if you want to continue to use a supported browser.

**Comment by xinuc** [comment]
@willsALMANJ hi, thanks for the releases.
out of curiosity, what do you use nowadays?

**Comment by wshanks** [comment]
@xinuc See the end of my first reply in this thread. Short version: VimFx, but I don't recommend it because it is also being abandoned. So far Vimium is the most promising option, but I will keep looking around to see if other things like cVim or Surfingkeys get ported over or new things developed like Saka-key or if the VimFx author ever makes a WebExtension (he said he would some day).

**Comment by mariusor** [comment]
In other news, [luakit](https://github.com/luakit/luakit) and [vimb](https://github.com/fanglingsu/vimb) are making a come back. Luakit was forked and has a new maintainer and vimb has it's original one continuing development.

They have been ported to work on top of webkit2gtk and gtk3, so I think they can become serious candidates. They are nowhere near the level of functionality that Firefox with pentadactyl offers, but for some people they might be enough.

**Comment by congma** [comment]
Again, many thanks to everyone for the input.

May I close this now, as we've found the answer to the question I posed?  Discussions about longer-term solutions could use another thread.

**Comment by pink-mist** [comment]
@congma it's your issue, so if you're satisfied it's been resolved, it makes sense for you to close it :)

Note that closing it will not prevent others from commenting, and it won't prevent @willsALMANJ from reopening it if he feels that's needed.

**Comment by wshanks** [comment]
Well, I'll close it since it seems like that's the consensus.

I still see one error in the error console about XPComShim that should probably be investigated, and there is a warning about Ary#itervalues being deprecated that should be kept in mind for Firefox 55 and 56 (not sure when the deprecation will change to an error).
